pkg/util/notification.go

```diff
@@ -38,4 +47,66 @@ const (
 
 	// AnnotationsPrefix is the notification annotations prefix
 	AnnotationsPrefix = "notifications." + configsync.GroupName
+
+	// SubscribeAnnotationPrefix is the annotation prefix of the notification subscription.
+	SubscribeAnnotationPrefix = AnnotationsPrefix + "/subscribe."
+
+	// MultiSubscriptionsField is the field name for subscriptions configured globally.
+	MultiSubscriptionsField = "subscriptions"
+
+	// NotifiedAnnotationPath is the annotation path that indicates it has been notified.
+	NotifiedAnnotationPath = ".metadata.annotations.notified." + AnnotationsPrefix
 )
+
+// NotificationEnabled returns whether the notification is enabled for the RSync object.
+func NotificationEnabled(ctx context.Context, client client.Client, rsNamespace string, annotations map[string]string, notificationConfig *v1beta1.NotificationConfig) (bool, error) {
+	hasSubscriptionAnnotation := false
+	// subscription via prefix
+	for key := range annotations {
+		if strings.HasPrefix(key, SubscribeAnnotationPrefix) {
+			hasSubscriptionAnnotation = true
+		}
+	}
+	if hasSubscriptionAnnotation && notificationConfig == nil {
+		return false, fmt.Errorf("subscription annotation was provided without a notificationConfig")
+	}
+	if notificationConfig == nil {
+		return false, nil
+	}
+	// ConfigMapRef must be defined
+	configMapName := v1beta1.GetConfigMapName(notificationConfig.ConfigMapRef)
+	if configMapName == "" {
+		return false, fmt.Errorf("notificationConfig must have a configMapRef")
+	}
+	// ensure ConfigMap exists
+	cm := &corev1.ConfigMap{}
+	cmObjectKey := types.NamespacedName{
+		Namespace: rsNamespace,
+		Name:      configMapName,
+	}
+	if err := client.Get(ctx, cmObjectKey, cm); err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, fmt.Errorf("notification ConfigMap %s not found in the %s namespace", cmObjectKey.Name, cmObjectKey.Namespace)
+		}
+		return false, err
+	}
+	// subscription via ConfigMap
+	_, found := cm.Data[MultiSubscriptionsField]
+	if !hasSubscriptionAnnotation && !found {
+		return false, fmt.Errorf("notificationConfig defined without any subscriptions")
+	}
```

Comment:
This should be allowed.
Suppose that a user originally enabled notification, but then the user just unsubscribed the annotation with the notification config left behind. It seems to be a valid case.

Reply:
That seems reasonable - I'd say it's valid but notifications are essentially disabled in that case. I decided to return an error here because I figured it may be confusing for users if they missed a step to fully configure notifications but there is no error.

Do you think we should return true or false?

Reply:
I feel like we should return false in this case. No subscription annotation, notificationConfig.configMapRef exists, but no subscription == subscription is disabled.

Otherwise, if a user wants to unsubscribe notification, he/she has to clean up the notificationConfig.

Reply:
done